clients/drcachesim/tests/tlb_simulator_unit_test.cpp

```diff
@@ -0,0 +1,187 @@
+
+    bool
+    process_memref(const memref_t &memref)
+    {
+        // Process memref like tlb_simulator does.
+        bool result = tlb_simulator_t::process_memref(memref);
+        // Save address used by tlb_simulator, if any.
+        if (!type_has_address(memref.data.type)) {
+            return result;
+        }
+        const memref_t *simref = &memref;
+        memref_t phys_memref;
+        if (knobs_.use_physical) {
+            phys_memref = memref2phys(memref);
+            simref = &phys_memref;
+        }
+        addresses.insert(simref->data.addr);
+        // Return the result of tlb_simulator.
+        return result;
+    };
+
+    void
+    set_knob_use_physical(bool set)
+    {
+        // XXX: We should really consider unifying the common knobs between simulator_t
+        // and its derived classes like tlb_simulator_t.
+        // Set tlb_simulator_t knob.
+        knobs_.use_physical = set;
+        // Set simulator_t knob.
+        knob_use_physical_ = set;
+    };
+
+    std::unordered_set<addr_t> addresses;
+};
+
+static memref_t
+generate_mem_ref(const addr_t addr, const addr_t pc)
+{
+    memref_t memref;
+    memref.data.type = TRACE_TYPE_READ;
+    memref.data.pid = 11111;
+    memref.data.tid = 22222;
+    memref.data.addr = addr;
+    memref.data.size = 8;
+    memref.data.pc = pc;
+    return memref;
+}
+
+// Checks that the addresses the tlb simulator uses (in memrefs), are the same we expect
+// in addresses.
+std::string
+check_addresses(const std::vector<memref_t> &memrefs,
+                const std::unordered_set<addr_t> &addresses,
+                const std::string &v2p_file_path = "")
+{
+    tlb_simulator_knobs_t knobs;
+    tlb_simulator_mock_t tlb_simulator_mock(knobs);
+
+    if (!v2p_file_path.empty()) {
+        tlb_simulator_mock.set_knob_use_physical(true);
+        std::ifstream fin;
+        fin.open(v2p_file_path);
+        if (!fin.is_open())
+            return "Failed to open the v2p file '" + v2p_file_path + "'\n";
+        std::string error_str = tlb_simulator_mock.create_v2p_from_file(fin);
+        fin.close();
+        if (!error_str.empty())
+            return "ERROR: v2p_reader failed with: " + error_str + "\n";
+    }
+
+    for (const memref_t &memref : memrefs)
+        tlb_simulator_mock.process_memref(memref);
+
+    if (addresses.size() != tlb_simulator_mock.addresses.size()) {
+        return "ERROR: size mismatch. Expected " + std::to_string(addresses.size()) +
+            " got " + std::to_string(tlb_simulator_mock.addresses.size()) + ".\n";
+    }
+
+    for (const addr_t addr : addresses) {
+        if (tlb_simulator_mock.addresses.count(addr) == 0)
+            return "ERROR: address " + std::to_string(addr) + " not found.\n";
+    }
+
+    return "";
+}
+
+void
+tlb_simulator_check_addresses(const char *testdir)
+{
+#ifdef X64
+    std::string v2p_file_path =
+        std::string(testdir) + "/drmemtrace.threadsig.aarch64.raw/raw/v2p.textproto";
```

Comment:
Should this not be in the raw/ directory?  raw/ is for data collected during tracing itself.

Reply:
Moved one dir up. Done.